core/src/main/java/com/nisovin/magicspells/spells/passive/HitArrowListener.java

```diff
@@ -0,0 +1,126 @@
+package com.nisovin.magicspells.spells.passive;
+
+import java.util.Set;
+import java.util.Map;
+import java.util.List;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.ArrayList;
+
+import org.bukkit.Material;
+import org.bukkit.entity.Arrow;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.inventory.ItemStack;
+
+import com.nisovin.magicspells.Spellbook;
+import com.nisovin.magicspells.MagicSpells;
+import com.nisovin.magicspells.spells.PassiveSpell;
+import com.nisovin.magicspells.util.OverridePriority;
+import com.nisovin.magicspells.util.magicitems.MagicItem;
+import com.nisovin.magicspells.util.magicitems.MagicItems;
+import com.nisovin.magicspells.util.magicitems.MagicItemData;
+
+public class HitArrowListener extends PassiveListener {
+	
+	private Set<Material> materials = new HashSet<>();
+	private Map<MagicItemData, List<PassiveSpell>> types = new HashMap<>();
+	private List<PassiveSpell> allTypes = new ArrayList<>();
+    
+    @Override
+    public void registerSpell(PassiveSpell spell, PassiveTrigger trigger, String var) {
+		if (var == null || var.isEmpty()) {
+			allTypes.add(spell);
+			return;
+		}
+		String[] split = var.split("\\|");
+		for (String s : split) {
+			s = s.trim();
+			MagicItem magicItem = MagicItems.getMagicItemFromString(s);
+			MagicItemData itemData = null;
+			if (magicItem != null) itemData = magicItem.getMagicItemData();
+			if (itemData == null) continue;
+
+			List<PassiveSpell> list = types.computeIfAbsent(itemData, material -> new ArrayList<>());
+			list.add(spell);
+			materials.add(itemData.getType());
+		}
+    }
+    
+    @OverridePriority
+    @EventHandler
+    public void onDamage(EntityDamageByEntityEvent event) {
+        Player player = getPlayerAttacker(event);
+        if (player == null || !(event.getEntity() instanceof LivingEntity)) {
+            return;
+        }
+        LivingEntity attacked = (LivingEntity) event.getEntity();
+        Spellbook spellbook = null;
+        
+        if (!allTypes.isEmpty()) {
+            spellbook = MagicSpells.getSpellbook(player);
+            for (PassiveSpell spell : allTypes) {
+                if (!isCancelStateOk(spell, event.isCancelled())) {
+                    continue;
+                }
+                if (!spellbook.hasSpell(spell, false)) {
+                    continue;
+                }
+                boolean casted = spell.activate(player, attacked);
+                if (!PassiveListener.cancelDefaultAction(spell, casted)) {
+                    continue;
+                }
+                event.setCancelled(true);
+            }
+        }
+        
+        if (!types.isEmpty()) {
```

Comment:
Since this is the last if, you should use `return` statements to prevent nested ifs.